src/sse/sse_utils.rs

```diff
@@ -0,0 +1,251 @@
+use core::arch::x86_64::*;
+
+//  __  __       _   _               _________  _     _ _
+// |  \/  | __ _| |_| |__           |___ /___ \| |__ (_) |_
+// | |\/| |/ _` | __| '_ \   _____    |_ \ __) | '_ \| | __|
+// | |  | | (_| | |_| | | | |_____|  ___) / __/| |_) | | |_
+// |_|  |_|\__,_|\__|_| |_|         |____/_____|_.__/|_|\__|
+//
+
+pub struct Rotate90F32 {
+    //sign_1st: __m128,
+    sign_2nd: __m128,
+    sign_both: __m128,
+}
+
+impl Rotate90F32 {
+    pub fn new(positive: bool) -> Self {
+        // There doesn't seem to be any need for rotating just the first element, but let's keep the code just in case
+        //let sign_1st = unsafe {
+        //    if positive {
+        //        _mm_set_ps(0.0, 0.0, 0.0, -0.0)
+        //    }
+        //    else {
+        //        _mm_set_ps(0.0, 0.0, -0.0, 0.0)
+        //    }
+        //};
+        let sign_2nd = unsafe {
+            if positive {
+                _mm_set_ps(0.0, -0.0, 0.0, 0.0)
+            } else {
+                _mm_set_ps(-0.0, 0.0, 0.0, 0.0)
+            }
+        };
+        let sign_both = unsafe {
+            if positive {
+                _mm_set_ps(0.0, -0.0, 0.0, -0.0)
+            } else {
+                _mm_set_ps(-0.0, 0.0, -0.0, 0.0)
+            }
+        };
+        Self {
+            //sign_1st,
+            sign_2nd,
+            sign_both,
+        }
+    }
+
+    #[inline(always)]
+    pub unsafe fn rotate_2nd(&self, values: __m128) -> __m128 {
+        let temp = _mm_shuffle_ps(values, values, 0xB4);
+        _mm_xor_ps(temp, self.sign_2nd)
+    }
+
+    // There doesn't seem to be any need for rotating just the first element, but let's keep the code just in case
+    //#[inline(always)]
+    //pub unsafe fn rotate_1st(&self, values: __m128) -> __m128 {
+    //    let temp = _mm_shuffle_ps(values, values, 0xE1);
+    //    _mm_xor_ps(temp, self.sign_1st)
+    //}
+
+    #[inline(always)]
+    pub unsafe fn rotate_both(&self, values: __m128) -> __m128 {
+        let temp = _mm_shuffle_ps(values, values, 0xB1);
+        _mm_xor_ps(temp, self.sign_both)
+    }
+}
+
+// Pack 1st complex
+// left: r1.re + r1.im, r2.re, r2.im
+// right: l1.re + l1.im, l2.re, l2.im
+// --> r1.re, r1.im, l1.re + l1.im
+#[inline(always)]
+pub unsafe fn pack_1st_f32(left: __m128, right: __m128) -> __m128 {
+    _mm_shuffle_ps(left, right, 0x44)
```

Comment:
If I'm understanding right, this is `_mm_unpacklo_pd()`. Note the `pd` instead of `ps`. If that's the case, I think this would be more clear if it more closely mirrored the `AvxVector` trait implementation for `__m128`:

```rust
 #[inline(always)]
    unsafe fn unpacklo_complex_f32(rows: [__m128; 2]) -> __m128 {
        let row0_double = _mm_castps_pd(rows[0]);
        let row1_double = _mm_castps_pd(rows[1]);
        let unpacked = _mm_unpacklo_pd(row0_double, row1_double);
        _mm_castpd_ps(unpacked)
    }
```

By making it as similar as possible, readers can transfer knowledge from the avx code set to sse and vice versa.

Reply:
Yes this seems to do the same. I can rename and use the same unpacklo (and unpackhi) instructions instead.

Reply:
I have switched to use the __unpacklo_pd instruction instead. 
Now I'm thinking about the function names. I have these, to do all the four possible shuffle variations:
- pack_1st_f32
- pack_2nd_f32
- pack_1and2_f32
- pack_2and1_f32
Not so happy with the "pack" verb, and "unpack" is also not great since they do a bit of both.. Repack, gather, extract..? The indices 1st and 2nd should probably be lo and hi instead.

Reply:
AVX256 has a very similar intrinsic to what this family of functions does, called "insertf128": https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=AVX&cats=Swizzle&expand=3150

Perhaps you could call it "insert_lo_lo()", "insert_lo_hi", "insert_hi_lo", etc

Reply:
Thinking about it more, I think I like extract instead of insert. Extract makes it sound more like you're pulling data out of the two source variables, and lo vs high says exactly which data you're pulling out of it.

Reply:
`extract_lo_lo(a, b)` etc is good!